// variant/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.strEnum = exports.constant = exports.exhaust = exports.property = exports.payload = exports.isType = exports.data = exports.fields = exports.variantModule = exports.variantFactory = exports.variantList = exports.variant = exports.types = exports.patterned = exports.outputTypes = exports.narrow = exports.flags = exports.keys = exports.keymap = exports.isOfVariant = exports.constrained = exports.cast = exports.augmented = void 0;
__exportStar(require("./deprecated"), exports);
var variant_1 = require("./variant");
Object.defineProperty(exports, "augmented", { enumerable: true, get: function () { return variant_1.augmented; } });
Object.defineProperty(exports, "cast", { enumerable: true, get: function () { return variant_1.cast; } });
Object.defineProperty(exports, "constrained", { enumerable: true, get: function () { return variant_1.constrained; } });
Object.defineProperty(exports, "isOfVariant", { enumerable: true, get: function () { return variant_1.isOfVariant; } });
Object.defineProperty(exports, "keymap", { enumerable: true, get: function () { return variant_1.keymap; } });
Object.defineProperty(exports, "keys", { enumerable: true, get: function () { return variant_1.keys; } });
Object.defineProperty(exports, "flags", { enumerable: true, get: function () { return variant_1.flags; } });
Object.defineProperty(exports, "narrow", { enumerable: true, get: function () { return variant_1.narrow; } });
Object.defineProperty(exports, "outputTypes", { enumerable: true, get: function () { return variant_1.outputTypes; } });
Object.defineProperty(exports, "patterned", { enumerable: true, get: function () { return variant_1.patterned; } });
Object.defineProperty(exports, "types", { enumerable: true, get: function () { return variant_1.types; } });
Object.defineProperty(exports, "variant", { enumerable: true, get: function () { return variant_1.variant; } });
Object.defineProperty(exports, "variantList", { enumerable: true, get: function () { return variant_1.variantList; } });
Object.defineProperty(exports, "variantFactory", { enumerable: true, get: function () { return variant_1.variantFactory; } });
Object.defineProperty(exports, "variantModule", { enumerable: true, get: function () { return variant_1.variantModule; } });
var tools_1 = require("./tools");
Object.defineProperty(exports, "fields", { enumerable: true, get: function () { return tools_1.fields; } });
Object.defineProperty(exports, "data", { enumerable: true, get: function () { return tools_1.data; } });
Object.defineProperty(exports, "isType", { enumerable: true, get: function () { return tools_1.isType; } });
Object.defineProperty(exports, "payload", { enumerable: true, get: function () { return tools_1.payload; } });
Object.defineProperty(exports, "property", { enumerable: true, get: function () { return tools_1.property; } });
Object.defineProperty(exports, "exhaust", { enumerable: true, get: function () { return tools_1.exhaust; } });
Object.defineProperty(exports, "constant", { enumerable: true, get: function () { return tools_1.constant; } });
var util_1 = require("./util");
Object.defineProperty(exports, "strEnum", { enumerable: true, get: function () { return util_1.strEnum; } });
__exportStar(require("./match"), exports);
__exportStar(require("./matcher"), exports);
__exportStar(require("./lookup"), exports);
__exportStar(require("./loose"), exports);
__exportStar(require("./generic"), exports);
__exportStar(require("./remote"), exports);
__exportStar(require("./sequence"), exports);
const variant_2 = __importDefault(require("./variant"));
exports.default = variant_2.default;
